### DISCUSSION TITLE
Why does a Server Side Component Store Data of an API Call in the rendered HTML?

### DISCUSSION THREAD

**Discussion by nihirv** [vercel/next.js#68238] [App Router]
I've been toying around with the /app router on a new project recently, and I'm calling an external API in my server component.

The entire return value of my API call seems to be stored in the rendered HTML. Specifically in some of the <script> tags that next.js itself is responsible for creating. These API calls have some information I don't want the client to be able to read, but currently it is directly in the DOM.

Here is high-level psuedo-code to show my implementation:
```javascript
const callAPI = (query: string) => {
  const res = await fetch(API_URL/query)
  return await res.json()
}

const MyComponent = () => {
  const query = "TEST"
  // results gets stringified and stored in the rendered HTML
  const results = await callAPI(query)
  return (
    // render components
  )
}
```

Any reason this happens? & can I prevent next from storing these API results in the rendered HTML?

**Comment by icyJoseph** [comment]
You need to read this:

- https://nextjs.org/blog/security-nextjs-server-components-actions

From there, this part is what you care most about right now:

```tsx
import Profile from './components/profile.tsx';
 
export async function Page({ params: { slug } }) {
  const [rows] = await sql`SELECT * FROM user WHERE slug = ${slug}`;
  const userData = rows[0];
  // EXPOSED: This exposes all the fields in userData to the client because
  // we are passing the data from the Server Component to the Client.
  // This is similar to returning `userData` in `getServerSideProps`
  return <Profile user={userData} />;
}
```

And the BAD, Profile implementation:

```tsx
'use client';
// BAD: This is a bad props interface because it accepts way more data than the
// Client Component needs and it encourages server components to pass all that
// data down. A better solution would be to accept a limited object with just
// the fields necessary for rendering the profile.
export default async function Profile({ user }: { user: User }) {
  return (
    <div>
      <h1>{user.name}</h1>
      ...
    </div>
  );
}
```

I suppose, you are passing your `const results = await callAPI(query)` down to a client component? Is that a correct assumption?

This could also be a problem in Pages Router, where you'd have to filter the data before passing it to the component, in the data-fetching-functions.
